src/cpp/src/block_manager.hpp

```diff
@@ -267,136 +520,285 @@ class BlockManager {
+                    }
+                    cached_blocks[hash] = last_blocks_vec;
+                }
             }
-        }
-
-        for (size_t i = 0; i < num_blocks; ++i) {
-
-            ov::genai::KVCacheBlock::Ptr block = nullptr; 
-            if (m_enable_prefix_caching) {
+            for (size_t i = 0; i < num_blocks; ++i) {
                 num_hashed_tokens += m_block_size;
                 if (num_hashed_tokens > content_length) {
                     num_hashed_tokens = content_length;
                 }
                 auto hash = sequence->get_hash(num_hashed_tokens);
-                block = m_allocator.allocate_block(hash, cached_blocks);
-            }
-            else {
-                block = m_allocator.allocate_block();
+                auto blocks_for_all_layers = m_allocator.allocate_block(hash, cached_blocks);
+                for (size_t layer_idx = 0; layer_idx < blocks_for_all_layers.size(); layer_idx++) {
+                    m_block_table[sequence_id][layer_idx].push_back(blocks_for_all_layers[layer_idx]);
+                }
             }
-            OPENVINO_ASSERT(block != nullptr);
-            m_block_table[sequence_id].push_back(block);
         }
     }
 
+    /**
+     * @return Percentage of KV cache used by all sequences.
+     */
     float get_used_percentage() const {
         return m_allocator.get_used_percentage();
     }
 
+    /**
+     * @brief Forks a sequence, establishing a new sequence from an existing one, reusing
+     * currently allocated blocks of the existing sequence.
+     * @param parent_id Parent sequence identifier
+     * @param child_id Sequence identifier for the new, forked sequence. Must be unique across
+     * other sequences tracked by this BlockManager.
+     */
     void fork_sequence(uint64_t parent_id, uint64_t child_id) {
         OPENVINO_ASSERT(m_block_table.count(child_id) == 0);
-        m_block_table[child_id].reserve(m_block_table[parent_id].size());
-        for (KVCacheBlock::Ptr & block : m_block_table[parent_id]) {
-            block->increment();
-            m_block_table[child_id].push_back(block);
+        m_block_table[child_id].resize(m_num_layers);
+        for (size_t layer_idx = 0; layer_idx < m_num_layers; layer_idx++) {
+            m_block_table[child_id][layer_idx].reserve(m_block_table[parent_id][layer_idx].size());
+            for (KVCacheBlock::Ptr &block: m_block_table[parent_id][layer_idx]) {
+                block->increment();
+                m_block_table[child_id][layer_idx].push_back(block);
+            }
         }
     }
 
+    /**
+     * @brief Frees all blocks for a given sequence.
+     * @param seq_id Identifier of the sequence to free.
+     */
     void free_sequence(size_t seq_id) {
-        auto block_table = m_block_table[seq_id];
-
-        for (KVCacheBlock::Ptr& block : block_table) {
-            m_allocator.free(block);
+        if (m_block_table.find(seq_id) == m_block_table.end()) {
+            return;
```

Comment:
why block_table can be empty? I suppose it's error in logic if we call free_sequence for non-existing ID

Reply:
Rewrote this as an assertion.